Add docs for customising node (#1221)
SeleniumHQ/seleniumhq.github.io: website_and_docs/content/documentation/grid/advanced_features/customize_node.en.md

---
title: "Customizing a Node"
linkTitle: "Customize Node"
weight: 4
aliases: [
"/documentation/en/grid/grid_4/customize_node/",
"/documentation/grid/advanced_features/customize_node/"
]
---

### How to customize a Node

> There are times when we would like a node to be customized to our needs. 

Reviewer: ```suggestion
There are times when we would like a Node to be customized to our needs. 
```

Author: Fixed

> For e.g., we may like to do some additional setup before a test case begins execution and some clean-up after a test case runs to completion.

Reviewer: ```suggestion
For e.g., we may like to do some additional setup before a session begins execution and some clean-up after a session runs to completion.
```

Author: Fixed

Following steps can be followed for this:

* Create a class that extends `org.openqa.selenium.grid.node.Node`
* Add a static method (this will be our factory method) to the newly created class whose signature looks like this: 

  `public static Node create(Config config)`. Here:

    * `Node` is of type `org.openqa.selenium.grid.node.Node`
    * `Config` is of type `org.openqa.selenium.grid.config.Config`
* Within this factory method, include logic for creating your new Class.
* To wire in this new customized logic into the hub, start the node and pass in the fully qualified class name of the above class to the argument `--node-implementation`

Let's see an example of all this:

1. Create a sample project using your favourite build tool (**Maven**|**Gradle**).
2. Add the below two dependencies to your sample project.
  * [org.seleniumhq.selenium/selenium-java](https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-java)

>   * [org.seleniumhq.selenium/selenium-grid](https://mvnrepository.com/artifact/org.seleniumhq.selenium/selenium-grid/4.6.0)

Reviewer: I believe only the Grid dependency is needed, isn't it?

Reviewer: Also, no need to have the URL with the exact version.

Author: @diemol - 

> I believe only the Grid dependency is needed, isn't it?

Yes and No to be honest. If one is building an uber/fat jar that one intends to use as the selenium server binary, then merely depending on `selenium-grid` will give only chrome capabilities. Only by depending on `selenium-java` will they be able to get support for the other browser flavors. I intend to send more PRs on this page which tackles both the use cases.

> Also, no need to have the URL with the exact version.

That was a typo :)

> 3. Add your customized node to the project.

Reviewer: ```suggestion
3. Add your customized Node to the project.
```

Author: Fixed

4. Build an [uber jar](https://imagej.net/develop/uber-jars) if you would like to just use `java -jar` to start the node.

> 5. Now start the node using the command:

Reviewer: ```suggestion
5. Now start the Node using the command:
```

```bash
java -jar selenium-server-4.6.0.jar node \

> --node-implementation org.seleniumhq.samples.CommentatingNode

Reviewer: You also need to add the uber jar to the class path, which can be done with the `--ext` flag. Like we do it in docker selenium https://github.com/SeleniumHQ/docker-selenium/blob/trunk/Hub/start-selenium-grid-hub.sh#L32-L34

Author: No. This is not needed. The jar name is what confused you. I will fix it. Essentially, if one builds the customised node into an uber jar, then there's no need to add `--ext`. The `--ext` is only used if one builds their implementation separately and then adds it along with the existing selenium-server jar (like how we added the tracing required jars and the `selenium-http-jdk-client.jar` instead of bundling everything into the same uber/fat jar.

Reviewer: You are right. Apologies.

```

> Here's a sample that just prints some messages on to the console whenever there's an activity of interest (session created, session deleted, a webdriver command executed etc.,) on the node.

Reviewer: ```suggestion
Here's a sample that just prints some messages on to the console whenever there's an activity of interest (session created, session deleted, a WebDriver command executed etc.,) on the Node.
```

Author: fixed

import org.openqa.selenium.grid.server.BaseServerOptions;
import org.openqa.selenium.internal.Either;
import org.openqa.selenium.remote.SessionId;
import org.openqa.selenium.remote.http.HttpRequest;
import org.openqa.selenium.remote.http.HttpResponse;
import org.openqa.selenium.remote.tracing.Tracer;

public class DecoratedLoggingNode extends Node {

  private Node node;

  protected DecoratedLoggingNode(Tracer tracer, URI uri, Secret registrationSecret) {
    super(tracer, new NodeId(UUID.randomUUID()), uri, registrationSecret);
  }

  public static Node create(Config config) {
    LoggingOptions loggingOptions = new LoggingOptions(config);
    BaseServerOptions serverOptions = new BaseServerOptions(config);
    URI uri = serverOptions.getExternalUri();
    SecretOptions secretOptions = new SecretOptions(config);
    Node node = LocalNodeFactory.create(config);
    DecoratedLoggingNode wrapper = new DecoratedLoggingNode(loggingOptions.getTracer(),
        uri, secretOptions.getRegistrationSecret());
    wrapper.node = node;
    return wrapper;
  }

  @Override
  public Either<WebDriverException, CreateSessionResponse> newSession(
      CreateSessionRequest sessionRequest) {

>     System.err.println("Before newSession()");

Reviewer: Any particular reason why `.err.` is used instead of `.out.`?

Author: No specific reason @diemol Just old habits that's all

Reviewer: I'd prefer to leave it with `out`, please.

Author: @diemol - Fixed. Amended the commit and force pushed it. Please help check now.
